### ISSUE TITLE
Function being called twice on LazyIterable findAny()

### ISSUE THREAD

**Issue by blacksteed232** [eclipse/eclipse-collections#994]
Hi,

I notice that with the following code, there is a duplicated function call.

```java
LazyIterable.collect(functions, func -> func.valueOf(someArg)).select(Objects::nonNull).getAny();
```
let say `functions` is an array of functions of this type:
```java
(someArg: String) -> {
  if ("CAR".equals(someArg)) return someArg;
 return null;
};
```

I notice that my function got called twice for the found value

This is not behaving like what I expected when I use `stream API`, for example:
```java
functions.stream().map(func -> func.valueOf(someArg)).filter(Objects::nonNull).findFirst().orElse(null);
```
here my function will only be called once for the found value.

Perhaps this has something to do with this [line](https://github.com/eclipse/eclipse-collections/blob/master/eclipse-collections/src/main/java/org/eclipse/collections/impl/lazy/CollectIterable.java#L116) in the codebase?

**Comment by prathasirisha** [comment]
Hi, thank you for bringing this to our attention. We are taking a look at it and will get back to you soon.

**Comment by donraab** [comment]
Hi @blacksteed232, thank you for raising this issue. This should be possible to replicate in a failing test. Looking at the code you linked to it certainly looks like the function will be executed twice. It sounds like @prathasirisha is looking into it so I'll wait to hear back what she finds.

As an aside, I believe the code you have written can be simplified as follows:

`LazyIterable.collect(functions, func -> func.valueOf(someArg)).detect(Objects::nonNull); `

It looks like this would result in the same issue.

**Comment by prathasirisha** [comment]
Hi @blacksteed232, we were able to replicate the issue and fixed it as part of this [PR.](https://github.com/eclipse/eclipse-collections/pull/998). This fix will be available in 11.0.0 release.

**Comment by donraab** [comment]
Hi @blacksteed232, I was able to validate the following test code works. Hopefully this is close enough to the use case you were seeing a failure on. I'm using a `MultiReaderList` to additionally guarantee we never call `iterator()`. A `MultiReaderList` will throw on any calls to `iterator()`.

```
@Test
public void testIssue994()
{
    AtomicInteger functionCount = new AtomicInteger(0);
    MultiReaderList<Integer> integers = Lists.multiReader.withAll(Interval.oneTo(5));
    Integer any = LazyIterate.collect(integers, functionCount::addAndGet).select(Objects::nonNull).getAny();
    Assert.assertEquals(1, any.longValue());
    Integer none = LazyIterate.collect(integers, functionCount::addAndGet).select(Objects::isNull).getAny();
    Assert.assertNull(none);
} 
```

**Comment by blacksteed232** [comment]
@prathasirisha @donraab Thank you so much! I will give it a try. Do you have a planned release cycle or an estimate on when v11 will come out?

**Comment by donraab** [comment]
Hi @blacksteed232, I'll leave it to @nikhilnanivadekar to answer, but my best guess would be at least 3-6 months as we just opened this major release for development. We just released 10.3 followed by 10.4 in rapid succession due to an issue we found in our JDK 15 EA build.

Do you need this fix urgently?

**Comment by blacksteed232** [comment]
@donraab no I don't need it urgently since I can use the alternative Stream API workaround, I just really like my codebase to be consistently using Eclipse collection :). Just curious though, is this considered a major breaking change that we have to bump a major version, instead of doing 10.5 or so?

**Comment by donraab** [comment]
@blacksteed232 We have already merged several breaking changes on the main branch so the next release version will be 11.0.0. We'd have to create a patch release to do a 10.5 (which we just did for 10.4). I'll leave that up to @nikhilnanivadekar as he handles our releases and each patch release is more work. A milestone release is much easier, and we usually will do these as we move along to test out changes. If we do a milestone release it would be named 11.0.0.M1.

**Comment by blacksteed232** [comment]
@donraab gotcha! Thank you for the explanation

**Comment by nikhilnanivadekar** [comment]
@blacksteed232 I can have the milestone version for you next week around Wednesday US time. If you would like a final version then 10.5.0 is an option but it’ll take me a couple of weeks to get to it. Let me know what will help your use case and happy to do either of the options.

**Comment by blacksteed232** [comment]
@nikhilnanivadekar thank you very much, it's not urgent at all so please do what is more convenient for you!

**Comment by nikhilnanivadekar** [comment]
@akphi sorry for the delay, I have released 11.0.0.M1 and it is available on Maven Central: https://search.maven.org/search?q=g:org.eclipse.collections%20AND%20v:11.0.0.M1
Please let us know if all issues are sorted out.
